### PR TITLE
Move manila cephfs spec definition under the related template

### PR DESCRIPTION
**Pull request by fmount**
Instead of hardcoding the `manila_cephfs` config to the `ansible` task,
this change organizes it into a template, that is rendered and applied
during the adoption run.
The template adds an initial support for `NFS` backend, in preparation to
a follow up patch where the `NFS` adoption test will be added.

### CODE REVIEW

**Review by gouthampacha — APPROVED** [review]
/lgtm 

thanks @fmount

**Comment by lkuchlan** [comment]
Should we create another file for Ceph-NFS ?

**Comment by fmount** [comment]
> Should we create another file for Ceph-NFS ?

Yes, but that can be done in a separate patch where we can focus on NFS testing.

**Comment by lkuchlan** [comment]
> > Should we create another file for Ceph-NFS ?
> 
> Yes, it's a good idea and that's exactly the purpose of the refactoring of this code. The NFS part can be done in a separate patch where we can focus on that part and test it locally as well.

Great! Thanks Francesco!
LGTM

**Review by lkuchlan — APPROVED** [review]

File: tests/roles/manila_adoption/files/manila_cephfs.yaml
Comment (fmount) [reply tests/roles/manila_adoption/files/manila_cephfs.yaml:42]: what we can do it moving this `manila_cephfs` into a template, and use the same file to render either native_cephfs or ceph_nfs.
For NFS we might need an additional information: the Ganesha old ganesha VIP in the ctlplane, and a couple of other things:

```
   customServiceConfig: |
            [DEFAULT]
            enabled_share_backends = cephfs
            host = hostgroup
            [cephfs]
            driver_handles_share_servers=False
            share_backend_name=tripleo_ceph
            share_driver=manila.share.drivers.cephfs.driver.CephFSDriver
            cephfs_conf_path=/etc/ceph/ceph.conf
            cephfs_auth_id=openstack
            cephfs_cluster_name=ceph
            cephfs_protocol_helper_type=NFS
            cephfs_nfs_cluster_id=cephfs
            cephfs_ganesha_server_ip=<VIP ON STORAGE NFS>
```
I'm going to move it to a template to support either native_cephfs or ceph_nfs. Still I think we should focus on NFS in a follow up.

File: tests/roles/manila_adoption/tasks/main.yaml
Comment (fmount) [inline tests/roles/manila_adoption/tasks/main.yaml:2]: Both `native_cephfs` and `cephnfs` use the same template, that is rendered with the appropriate config and applied through `oc patch ..`

File: tests/roles/manila_adoption/tasks/main.yaml
Comment (fmount) [inline tests/roles/manila_adoption/tasks/main.yaml:7]: @jistr I'm not sure if there's a preference to render the template elsewhere or it's ok to use /tmp/

File: tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2
Comment (fmount) [inline tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2:51]: @lkuchlan FYI +1 on ternary

File: tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2
Comment (fmount) [inline tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2:37]: @gouthampacha I assume it's ok to not have `enabled_share_protocols` in the `share` config.

**Comment by fmount** [comment]
@gouthampacha @lkuchlan I think I addressed the request to have a template instead of a file, and render it according to cephfs/nfs.
Let's see the CI in action to gather the very first results.

File: tests/roles/manila_adoption/files/manila_cephfs.yaml
Comment (lkuchlan) [reply tests/roles/manila_adoption/files/manila_cephfs.yaml]: Yes, that is another possibility. Sounds good to me.

Comment (fmount) [reply tests/roles/manila_adoption/files/manila_cephfs.yaml]: this is what I implemented in the current patch set, let's see if CI is happy and we can build the NFS use case on top of this.

File: tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2
Comment (gouthampacha) [reply tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2:37]: ack; makes sense..

File: tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2
Comment (gouthampacha) [inline tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2:40]: the host must be set to "hostgroup" on both drivers..

File: tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2
Comment (gouthampacha) [inline tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2:53]: typo in "default"

Comment (fmount) [reply tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2:53]: thanks, good catch!

File: tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2
Comment (fmount) [reply tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2:40]: ack, thanks, good catch, we'll fix it also in [1][2]

[1] https://github.com/openstack-k8s-operators/manila-operator/blob/main/config/samples/backends/ceph-nfs/backend.yaml
[2] https://github.com/openstack-k8s-operators/manila-operator/blob/main/config/samples/backends/cephfs/backend.yaml

Comment (fmount) [reply tests/roles/manila_adoption/templates/manila_cephfs.yaml.j2]: Fixed for now here.

**Review by gouthampacha — APPROVED** [review]
/lgtm

Thanks @fmount

**Comment by fmount** [comment]
recheck

**Comment by fmount** [comment]
recheck

**Comment by fmount** [comment]
recheck

File: tests/roles/manila_adoption/tasks/main.yaml
Comment (jistr) [reply tests/roles/manila_adoption/tasks/main.yaml:7]: Elsewhere we've used tmp dir relative to the ansible's working directory, which is the playbook directory:

https://github.com/openstack-k8s-operators/data-plane-adoption/blob/631d249e5fe47273abfd8149f9fb4fd625dcd853/tests/roles/mariadb_copy/defaults/main.yaml#L2

So in that example it becomes `<repo-dir>/playbooks/tmp/mariadb` which is slightly easier to work with manually than digging through the content of `/tmp` i think, but it's not a blocker.

**Review by jistr — APPROVED** [review]

**Comment by fmount** [comment]
From [1] I can see:

```
TASK [manila_adoption : deploy podified Manila with cephfs backend] ************
changed: [localhost] => {"changed": true, "cmd": "set -euxo pipefail\n\n\noc patch openstackcontrolplane openstack --type=merge --patch-file=/tmp/manila_cephfs.yaml\n", "delta": "0:00:00.319010", "end": "2024-08-06 03:57:54.882576", "msg": "", "rc": 0, "start": "2024-08-06 03:57:54.563566", "stderr": "+ oc patch openstackcontrolplane openstack --type=merge --patch-file=/tmp/manila_cephfs.yaml", "stderr_lines": ["+ oc patch openstackcontrolplane openstack --type=merge --patch-file=/tmp/manila_cephfs.yaml"], "stdout": "openstackcontrolplane.core.openstack.org/openstack patched", "stdout_lines": ["openstackcontrolplane.core.openstack.org/openstack patched"]}

TASK [manila_adoption : wait for Manila to start up] ***************************
FAILED - RETRYING: [localhost]: wait for Manila to start up (60 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (59 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (58 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (57 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (56 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (55 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (54 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (53 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (52 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (51 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (50 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (49 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (48 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (47 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (46 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (45 retries left).
changed: [localhost] => {"attempts": 17, "changed": true, "cmd": "set -euxo pipefail\n\n\noc get pod --selector=component=manila-api -o jsonpath='{.items[0].status.phase}{\"\\n\"}' | grep Running\noc get pod --selector=component=manila-scheduler -o jsonpath='{.items[0].status.phase}{\"\\n\"}' | grep Running\noc get pod --selector=component=manila-share -o jsonpath='{.items[0].status.phase}{\"\\n\"}' | grep Running\n", "delta": "0:00:00.678800", "end": "2024-08-06 03:58:36.966862", "msg": "", "rc": 0, "start": "2024-08-06 03:58:36.288062", "stderr": "+ oc get pod --selector=component=manila-api -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'\n+ grep Running\n+ oc get pod --selector=component=manila-scheduler -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'\n+ grep Running\n+ oc get pod --selector=component=manila-share -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'\n+ grep Running", "stderr_lines": ["+ oc get pod --selector=component=manila-api -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'", "+ grep Running", "+ oc get pod --selector=component=manila-scheduler -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'", "+ grep Running", "+ oc get pod --selector=component=manila-share -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'", "+ grep Running"], "stdout": "Running\nRunning\nRunning", "stdout_lines": ["Running", "Running", "Running"]}

TASK [manila_adoption : check that Manila is reachable and its endpoints are defined] ***
RETRYING: [localhost]: check that Manila is reachable and its endpoints are defined (15 retries left).
changed: [localhost] => {"attempts": 2, "changed": true, "cmd": "set -euxo pipefail\n\n\nalias openstack=\"oc exec -t openstackclient -- openstack\"\n\n${BASH_ALIASES[openstack]} endpoint list | grep -i share\n${BASH_ALIASES[openstack]} share pool list\n", "delta": "0:00:07.721853", "end": "2024-08-06 03:58:55.041623", "msg": "", "rc": 0, "start": "2024-08-06 03:58:47.319770", "stderr": "+ alias 'openstack=oc exec -t openstackclient -- openstack'\n+ oc exec -t openstackclient -- openstack endpoint list\n+ grep -i share\n+ oc exec -t openstackclient -- openstack share pool list", "stderr_lines": ["+ alias 'openstack=oc exec -t openstackclient -- openstack'", "+ oc exec -t openstackclient -- openstack endpoint list", "+ grep -i share", "+ oc exec -t openstackclient -- openstack share pool list"], "stdout": "| 01cf992ee5d54258910a43326f8bbb12 | regionOne | manilav2     | sharev2        | True    | internal  | http://manila-internal.openstack.svc:8786/v2                        |\n| 8be5443cadcc450f8b421c166ae387e6 | regionOne | manilav2     | sharev2        | True    | public    | https://manila-public-openstack.apps-crc.testing/v2                 |\n| be645d07a39b456794ddeb0f66250484 | regionOne | manila       | share          | True    | public    | https://manila-public-openstack.apps-crc.testing/v1/%(project_id)s  |\n| f671dd69485e49e1a83aef2b4e00115a | regionOne | manila       | share          | True    | internal  | http://manila-internal.openstack.svc:8786/v1/%(project_id)s         |\n+-------------------------+-----------+---------+--------+\n| Name                    | Host      | Backend | Pool   |\n+-------------------------+-----------+---------+--------+\n| hostgroup@cephfs#cephfs | hostgroup | cephfs  | cephfs |\n+-------------------------+-----------+---------+--------+", "stdout_lines": ["| 01cf992ee5d54258910a43326f8bbb12 | regionOne | manilav2     | sharev2        | True    | internal  | http://manila-internal.openstack.svc:8786/v2                        |", "| 8be5443cadcc450f8b421c166ae387e6 | regionOne | manilav2     | sharev2        | True    | public    | https://manila-public-openstack.apps-crc.testing/v2                 |", "| be645d07a39b456794ddeb0f66250484 | regionOne | manila       | share          | True    | public    | https://manila-public-openstack.apps-crc.testing/v1/%(project_id)s  |", "| f671dd69485e49e1a83aef2b4e00115a | regionOne | manila       | share          | True    | internal  | http://manila-internal.openstack.svc:8786/v1/%(project_id)s         |", "+-------------------------+-----------+---------+--------+", "| Name                    | Host      | Backend | Pool   |", "+-------------------------+-----------+---------+--------+", "| hostgroup@cephfs#cephfs | hostgroup | cephfs  | cephfs |", "+-------------------------+-----------+---------+--------+"]}
```

so I think we're good to go here.
Also, as per the must-gather [2] manilaShare (share1) looks like:

```
      manilaShares:
        share1:
          customServiceConfig: |
            [DEFAULT]
            enabled_share_backends=cephfs
            host = hostgroup

            [cephfs]
            driver_handles_share_servers=False
            share_backend_name=cephfs
            share_driver=manila.share.drivers.cephfs.driver.CephFSDriver
            cephfs_conf_path=/etc/ceph/ceph.conf
            cephfs_auth_id=openstack
            cephfs_cluster_name=ceph
            cephfs_volume_mode=0755
            cephfs_protocol_helper_type= CEPHFS
          networkAttachments:
          - storage
          replicas: 1
```

which is the expected configuration from this job. (/cc @gouthampacha @jistr @lkuchlan).

[1] https://logserver.rdoproject.org/14/514/995c8fefeb1ce585dde4a9a577eb1508036fc617/github-check/adoption-standalone-to-crc-ceph/012d62b/controller/data-plane-adoption-tests-repo/data-plane-adoption/tests/logs/test_with_ceph_out_2024-08-06T03:34:14EDT.log
[2] https://logserver.rdoproject.org/14/514/995c8fefeb1ce585dde4a9a577eb1508036fc617/github-check/adoption-standalone-to-crc-ceph/012d62b/controller/ci-framework-data/logs/openstack-k8s-operators-openstack-must-gather/namespaces/openstack/crs/openstackcontrolplanes.core.openstack.org/openstack.yaml

**Review by gouthampacha — APPROVED** [review]
Thanks @fmount

**Comment by lkuchlan** [comment]
Thanks LGTM

**Comment by fmount** [comment]
rebased

**Comment by fmount** [comment]
I suspect we need to wait for @jistr to land this one, unless someone else can help w/ reviews (/cc @cescgina || @marios ).
FYI we had green run and the template was properly rendered and applied.
The next step will be start working on the NFS scenario that does not exist yet.

**Comment by fmount** [comment]
rebased

File: tests/roles/manila_adoption/tasks/main.yaml
Comment (marios) [inline tests/roles/manila_adoption/tasks/main.yaml:7]: just wondering if it would be useful to have that somewhere other than /tmp so we can get it in logs for debug

**Review by marios — COMMENTED** [review]
lgtm and should be noop for the existing tests i.e. uses cephfs backend
job looks happy

https://logserver.rdoproject.org/14/514/dee4c9ee49de0c4fae73834498c4fb8bd78f8fcf/github-check/adoption-standalone-to-crc-ceph/bccd973/controller/data-plane-adoption-tests-repo/data-plane-adoption/tests/logs/test_with_ceph_out_2024-08-12T04:19:10EDT.log

```

TASK [manila_adoption : deploy podified Manila with cephfs backend] ************
changed: [localhost] => {"changed": true, "cmd": "set -euxo pipefail\n\n\noc patch openstackcontrolplane openstack --type=merge --patch-file=/tmp/manila_cephfs.yaml\n", "delta": "0:00:00.304621", "end": "2024-08-12 04:43:48.507795", "msg": "", "rc": 0, "start": "2024-08-12 04:43:48.203174", "stderr": "+ oc patch openstackcontrolplane openstack --type=merge --patch-file=/tmp/manila_cephfs.yaml", "stderr_lines": ["+ oc patch openstackcontrolplane openstack --type=merge --patch-file=/tmp/manila_cephfs.yaml"], "stdout": "openstackcontrolplane.core.openstack.org/openstack patched", "stdout_lines": ["openstackcontrolplane.core.openstack.org/openstack patched"]}

TASK [manila_adoption : wait for Manila to start up] ***************************
FAILED - RETRYING: [localhost]: wait for Manila to start up (60 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (59 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (58 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (57 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (56 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (55 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (54 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (53 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (52 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (51 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (50 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (49 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (48 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (47 retries left).
FAILED - RETRYING: [localhost]: wait for Manila to start up (46 retries left).
changed: [localhost] => {"attempts": 16, "changed": true, "cmd": "set -euxo pipefail\n\n\noc get pod --selector=component=manila-api -o jsonpath='{.items[0].status.phase}{\"\\n\"}' | grep Running\noc get pod --selector=component=manila-scheduler -o jsonpath='{.items[0].status.phase}{\"\\n\"}' | grep Running\noc get pod --selector=component=manila-share -o jsonpath='{.items[0].status.phase}{\"\\n\"}' | grep Running\n", "delta": "0:00:00.563419", "end": "2024-08-12 04:44:26.994515", "msg": "", "rc": 0, "start": "2024-08-12 04:44:26.431096", "stderr": "+ oc get pod --selector=component=manila-api -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'\n+ grep Running\n+ oc get pod --selector=component=manila-scheduler -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'\n+ grep Running\n+ oc get pod --selector=component=manila-share -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'\n+ grep Running", "stderr_lines": ["+ oc get pod --selector=component=manila-api -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'", "+ grep Running", "+ oc get pod --selector=component=manila-scheduler -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'", "+ grep Running", "+ oc get pod --selector=component=manila-share -o 'jsonpath={.items[0].status.phase}{\"\\n\"}'", "+ grep Running"], "stdout": "Running\nRunning\nRunning", "stdout_lines": ["Running", "Running", "Running"]}
```

**Review by marios — APPROVED** [review]
forgot to select approve :)

**Comment by gouthampacha** [comment]
@marios thank you :) but, the label didn't get applied.. maybe you need an explicit `/lgtm` or `/approve`?

File: tests/roles/manila_adoption/tasks/main.yaml
Comment (fmount) [reply tests/roles/manila_adoption/tasks/main.yaml:7]: I was discussing a similar thing earlier in the patch with @jistr, but in general this template is rendered and applied, which means I will find the result of what is applied in the main manila CR collected by must-gather. However I think it would be useful, as a future improvement, to put this resulting template elsewhere (maybe a path we collect in CI logs).
Let me know if you have thoughts or we did something similar for other components and I'll follow the pattern.
I might follow the same pattern I saw here [1] and render templates under `../config/`.

[1] https://github.com/openstack-k8s-operators/data-plane-adoption/blob/main/tests/roles/backend_services/tasks/main.yaml#L64

**Pull request merged**
